### DISCUSSION TITLE
SPI issue on Jetson nano

### DISCUSSION THREAD

**Discussion by Sarim-Jalil** [OE4T/meta-tegra#1723] [Q&A] [answered]
Hi, I have read the [https://github.com/OE4T/meta-tegra/issues/573](url) discussion and worked as described in it.

I am working with jetson nano and I am using branch Kirkston-lt4-r32.7.x
I have done all the settings mentioned in the discussion and flashed the jetson nano. My pins are coming up as spi1 and I have spidev enabled but when I try to do the spi loopback test it doesnot pass.

root@jetson-nano-devkit:/usr/bin# spiTest -v -D /dev/spidev0.0
spi mode: 0x0
bits per word: 8
max speed: 500000 Hz (500 KHz)
TX | FF FF FF FF FF FF 40 00 00 00 00 95 FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF F0 0D | ......@....�..................�.
RX | 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 | ................................

Below is my patch file and pinctrl outputs
[pinctrl.txt](https://github.com/user-attachments/files/17363135/pinctrl.txt)
[patch.txt](https://github.com/user-attachments/files/17363136/patch.txt)

The only difference I could see is the Status = default as compare to Status = 2 in your pinctrl file.

Below is the procedure I have followed.

1- Made the patch file using devtool.
2- Enabled spidev from menuconfig.
3- Applied patch file and Built the image.
4- Flashed the dev-kit and sd-card using -b00.
5- Shorted pin 19 and 21 on my dekit.
6- Run the loopback test.

Kindly let me know if you need any other info or command run on the dev-kit.

**Comment by Sarim-Jalil** [comment]
@madisongh sorry for disturbing you but I need this a bit urgent.

**Comment by madisongh** [comment, marked as answer]
You missed removing the `TEGRA_GPIO(C,` pins from the gpio controller in your patch. Looks like that's still necessary with 32.7.5, I couldn't even get SPI working with the stock Jetson Linux without it. The attached patch includes the missing bits, and covers both -b00 and -a02 carriers. It also removes the SPI2 pins from the gpio controller, although it doesn't configure the function in the pinmux config.

[0001-nvidia-platform-t210-configure-spi1-pins-on-40-pin-h.patch.txt](https://github.com/user-attachments/files/17450601/0001-nvidia-platform-t210-configure-spi1-pins-on-40-pin-h.patch.txt)

Here's the `spidev_test` output with this patch:
```
root@jetson-nano-devkit:~# modprobe spidev
root@jetson-nano-devkit:~# spidev_test -v -D /dev/spidev0.0 
spi mode: 0x0
bits per word: 8
max speed: 500000 Hz (500 KHz)
TX | FF FF FF FF FF FF 40 00 00 00 00 95 FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF F0 0D  | ......@....�..................�.
RX | FF FF FF FF FF FF 40 00 00 00 00 95 FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF FF F0 0D  | ......@....�..................�.
```

**Reply by Sarim-Jalil** [reply]
Thank you so much. GPIO usage was indeed the issue and you are right you also have to flash the dev board in order to make it work otherwise only flashing the SD card wont work. One other thing is using the FAB parameter as by default the dtb selected in the scripts is for 2gb dev board.

**Comment by abinaya2003-s** [comment]
Applied your patch that sets the spi1_mosi/miso/sck/cs0/cs1 pinmux functions to spi1 (from rsvd1) in tegra210-porg-pinmux-p3448-0000-b00.dtsi.
Enabled spidev in kernel config (CONFIG_SPI_SPIDEV=m).
Rebuilt and flashed the image (using -b00 for the B00 carrier).
Changed extlinux.conf to use /boot/tegra210-p3448-0000-p3449-0000-b00.dtb instead of the custom A02 DTB.
Modified the DTB file itself: I decompiled /boot/tegra210-p3448-0000-p3449-0000-b00.dtb using dtc, removed the SPI1 pin hex values (0xbc, 0xbd, 0xbe, 0xc1) from the gpio-input property in the default node, recompiled it, and replaced the original file.
Rebooted and verified:
/dev/spidev0.0, /dev/spidev0.1, /dev/spidev1.0, /dev/spidev1.1 all exist.
sudo cat /sys/kernel/debug/tegra_gpio | grep -i spi1 returns nothing (SPI pins no longer claimed by GPIO).
The pinmux shows function spi1 for the relevant pins (checked via pinctrl debugfs).  


Problem:

I shorted Pin 19 (MOSI) to Pin 21 (MISO) and ran:
bash

sudo ./spidev_test -D /dev/spidev0.0 -v -p "1234"

Output:
text

TX | 31 32 33 34 ... | 1234
RX | 00 00 00 00 ... | ....

All SPI devices (0.0, 0.1, 1.0, 1.1) give the same result. I also tried all SPI modes (0,1,2,3) but still zeros.
Questions:

    What else could be wrong? Is it possible the spidev driver is not binding to the correct node? I checked /proc/device-tree/spi@7000d400/spi@0/compatible and it shows "tegra-spidev" – is that correct?

    Could there be a hardware issue? I verified the jumper is on pins 19 and 21 – maybe I should try a different loopback (like SCLK→MISO) to test the receiver? If so, which pins?

    Is there a missing patch for the spi@7000d400 controller node itself (e.g., status = "okay" or spi@0 child node)?

    Should I check the gpio-output or gpio-input more carefully? I removed 0xbc,0xbd,0xbe,0xc1 but maybe 0xc2 (CS1) also needs removal? Or perhaps there's another property affecting the SPI pins?

I'd appreciate any guidance you can provide. The TX is fine, but I'm stuck on the RX side.

Thanks!
